### PR TITLE
Add storage and module API methods to get monthly active users and their appservices

### PR DESCRIPTION
**Pull request by buffless-matt**
### Description

This PR adds a storage method that is similar to the pre-existing [get_monthly_active_count_by_service](https://github.com/matrix-org/synapse/blob/28199e93579b5a73841a95ed4d355322227432b5/synapse/storage/databases/main/monthly_active_users.py#L96-L123), but different in the following ways:

- The new method does not group and count users under appservices. I.e. It returns each active user individually.
- The new method has optional timestamp filtration capability.

There is also an added module API method that plumbs into the new storage method (so that it is accessible via the module API).

The purpose of this PR is to be able to identify individual active users (and not just counts of grouped users). There are more details (about the purpose) available to those who have permission to view the linked issue/comment below.

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

File: synapse/storage/databases/main/monthly_active_users.py [hunk 1]
```diff
@@ -122,6 +122,49 @@ def _count_users_by_service(txn: LoggingTransaction) -> Dict[str, int]:
...
+        elif end_timestamp is not None:
+            where_clause = 'WHERE "timestamp" <= ?'
+            query_params = [end_timestamp]
+        else:
+            where_clause = ""
+            query_params = []
+
+        def _list_users(txn: LoggingTransaction) -> List[Tuple[str, str]]:
+            sql = f"""
+                    SELECT COALESCE(appservice_id, 'native'), user_id
```
Comment (erikjohnston) [inline synapse/storage/databases/main/monthly_active_users.py:157]: Can we just return `null` if there isn't an appservice? It saves having to use a magic string

File: synapse/module_api/__init__.py [hunk 2]
```diff
@@ -1419,6 +1419,26 @@ async def set_push_rule_action(
...
+        Please see corresponding storage docstring for more details.
+
+        Arguments:
+            start_timestamp: If specified, only include users that were first active
+                at or after this point
+            end_timestamp: If specified, only include users that were first active
+                at or before this point
+
+        Returns:
+            A list of tuples (appservice_id, user_id)
```
Comment (erikjohnston) [inline synapse/module_api/__init__.py:1435]: Can you document what is returned for `appservice_id` if the user isn't from an appservice? (As per below I think we should return `None` in that case)

File: synapse/storage/databases/main/monthly_active_users.py (hunk 1 above)
Comment (buffless-matt) [reply synapse/storage/databases/main/monthly_active_users.py:157]: We could live with that, but the pre-existing [get_monthly_active_count_by_service](https://github.com/matrix-org/synapse/blob/28199e93579b5a73841a95ed4d355322227432b5/synapse/storage/databases/main/monthly_active_users.py#L111) method (which this new method is modelled on) is emitting `"native"`.

File: synapse/module_api/__init__.py (hunk 2 above)
Comment (buffless-matt) [reply synapse/module_api/__init__.py:1435]: I was hoping [this](https://github.com/buffless-matt/synapse/blob/451d7becda6c3e4971771bf12bce6aaba27a2315/synapse/storage/databases/main/monthly_active_users.py#L129-L130) sentence would alleviate the need to repeat what is mentioned in `get_monthly_active_count_by_service`, but for the sake of reducing the coupling between the two docstrings, I've added it here: 95accd8

File: synapse/storage/databases/main/monthly_active_users.py (hunk 1 above)
Comment (erikjohnston) [reply synapse/storage/databases/main/monthly_active_users.py:159]: Oh interesting, fair enough then. I think it probably does make more sense to be consistent then

File: synapse/module_api/__init__.py (hunk 2 above)
Comment (erikjohnston) [reply synapse/module_api/__init__.py:1435]: Ah, I see. I think the docstring should either document the return values fully, or not at all.

**Review by erikjohnston — APPROVED** [review]

**Comment by buffless-matt** [comment]
@erikjohnston thank you for the review (I appreciate it).

What needs to happen now to get this merged?

I notice there are some CI failures, but I don't think they are my doing. The failures appear unrelated to my patch and it seems like other PRs are having similar issues.

**Comment by erikjohnston** [comment]
@buffless-matt Hmm, CI failed and so the auto-merge didn't trigger. The CI failure looked unrelated to this PR, so I've merged in  develop to see if that was the issue.

**Pull request merged**